### ISSUE TITLE
Provide a FORCE_HTTPS option

### ISSUE THREAD

**Issue by CAFxX** [cloudfoundry/php-buildpack#255]
What version of Cloud Foundry and CF CLI are you using? (i.e. What is the output of running `cf curl /v2/info && cf version`?
capi 2.95.0

What version of the buildpack you are using?
master

If you were attempting to accomplish a task, what was it you were attempting to do?
Force redirect HTTP to HTTPS

What did you expect to happen?
Ideally, a mechanism similar to the one provided by the staticfile buildpack (FORCE_HTTPS). As an example, a FORCE_HTTPS option in `options.json` that automatically sets the relevant configuration in either apache or nginx would be the ideal outcome.

What was the actual behavior?
As a user, I have to know specifics about SSL termination done by my platform operator and I have to write custom httpd/nginx configuration

**Comment by dmikusa-pivotal** [comment]
It could be a convenience.  You can do it at the app level w/PHP or you can do it by including a HTTPD/Nginx config snippet.  You don't need to override the entire config though, so it's fairly low impact to do this way.

https://docs.cloudfoundry.org/buildpacks/php/gsg-php-config.html#engine-configurations

**Comment by CAFxX** [comment]
@dmikusa-pivotal to do it at the httpd/nginx config level you basically have to override the whole configuration, since there is no "wildcard include" feature AFAIK. Even the docs say this, although in a very roundabout way:

> You can add your own configuration files. The components will not know about these, so you must ensure that they are included. For example, you can add an include directive to the httpd configuration to include your file:
> ```
> ServerRoot "${HOME}/httpd"
> Listen ${PORT}
> ServerAdmin "${HTTPD_SERVER_ADMIN}"
> ServerName "0.0.0.0"
> DocumentRoot "${HOME}/#{WEBDIR}"
> Include conf/extra/httpd-modules.conf
> Include conf/extra/httpd-directories.conf
> Include conf/extra/httpd-mime.conf
> Include conf/extra/httpd-logging.conf
> Include conf/extra/httpd-mpm.conf
> Include conf/extra/httpd-default.conf
> Include conf/extra/httpd-remoteip.conf
> Include conf/extra/httpd-php.conf
> Include conf/extra/httpd-my-special-config.conf # This line includes your additional file.
> ```

This means the user has to override the "root" apache configuration to be able to add custom config. If a newer version of the buildpack provides a new version of the root configuration then, likely, the user won't notice and will run with a broken configuration.

And if you do it at the PHP level it only applies to requests handled by PHP (so static assets served directly by nginx/apache are not covered).

To summarize: yes, it can be done in other ways but they are unergonomic and prone to cause issues down the road.

**Comment by dmikusa-pivotal** [comment]
While a wildcard feature would be convenient (@sclevine - this could be a worthwhile improvement, for this and other reasons), it's not totally necessary.  You can take any one of the existing config files and append additional config to it.  These files are quite unlikely to change, so you could pick one like [httpd-directories](https://github.com/cloudfoundry/php-buildpack/blob/master/defaults/config/httpd/extra/httpd-directories.conf) that hasn't changed in three years and use that.

Ex:  take the default & append the redirect to the end of it.

```
<Directory />
    AllowOverride none
    Require all denied
</Directory>

<Directory "${HOME}/#{WEBDIR}">
    Options SymLinksIfOwnerMatch
    AllowOverride All
    Require all granted
</Directory>

<Files ".ht*">
    Require all denied
</Files>

RewriteEngine On
RewriteCond %{HTTPS} !=on
RewriteCond %{HTTP:X-Forwarded-Proto} !https [NC]
RewriteRule ^ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]
```

You can do something similar for Nginx, the config changes quite infrequently.

**Comment by dmikusa-pivotal** [comment]
Oh, the other thing you can do to make this work is add a `.htaccess` file to the root of your `htdocs` folder.  This will let you run certain bits of HTTPD config, including rewrite rules so you can drop the following bit into a `.htaccess` file to redirect from HTTP to HTTPS

```
RewriteEngine On
RewriteCond %{HTTPS} !=on
RewriteCond %{HTTP:X-Forwarded-Proto} !https [NC]
RewriteRule ^ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]
```

**Comment by sclevine** [comment]
@CAFxX Does @dmikusa-pivotal's `.htaccess` solution provide all of the functionality you need?

**Comment by CAFxX** [comment]
@sclevine sorry for the late response. That could work for apache, but if the user chooses nginx instead it is left out in the cold. 

Also, users are often confused by why certain buildpacks (staticfile) provide this functionality built-in, whereas others don't. From a ux pespective having a single way of requesting this behavior would be preferable.

**Comment by sclevine** [comment]
The staticfile and PHP buildpacks provide a dedicated webserver. It wouldn't be possible to add this functionality to the rest of the buildpacks without adding webservers to all of them.

In the staticfile case, it isn't possible to use application code, so we offer `FORCE_HTTPS`.

We are considering making some major changes to this buildpack after merging #260. If this feature is still applicable after that, we will definitely consider adding it.

**Comment by dfreilich** [comment]
@sclevine @dmikusa-pivotal I assume that this isn't going to be done for the php-buildpack, but is this relevant for the php-cnb? Should I port the issue there?
